Core/src/Geometry/ProtoDetector.cpp

```diff
@@ -0,0 +1,210 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2018-2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Geometry/ProtoDetector.hpp"
+
+#include "Acts/Utilities/Enumerate.hpp"
+
+#include <exception>
+#include <sstream>
+
+void Acts::ProtoVolume::extendUp(Acts::ProtoVolume& ptVolume) {
+  ptVolume.extent.extend(extent);
+
+  for (auto& cv : constituentVolumes) {
+    ptVolume.extent.extend(cv.extent);
+    cv.extendUp(ptVolume);
+  }
+}
+
+void Acts::ProtoVolume::propagateMinDown(BinningValue bValue) {
+  for (auto& cv : constituentVolumes) {
+    cv.extent.set(bValue, extent.min(bValue), cv.extent.max(bValue));
+    cv.propagateMinDown(bValue);
+  }
+}
+
+void Acts::ProtoVolume::propagateMaxDown(BinningValue bValue) {
+  for (auto& cv : constituentVolumes) {
+    cv.extent.set(bValue, cv.extent.min(bValue), extent.max(bValue));
+    cv.propagateMaxDown(bValue);
+  }
+}
+
+void Acts::ProtoVolume::constrainDown(const Acts::ProtoVolume& ptVolume) {
+  extent.addConstrain(ptVolume.extent);
+  for (auto& cv : constituentVolumes) {
+    cv.extent.addConstrain(extent);
+  }
+}
+
+void Acts::ProtoVolume::harmonize(bool legacy) {
+  std::vector<BinningValue> otherConstrains;
+
+  // Deal with the constituents
+  if (not constituentVolumes.empty()) {
+    if (constituentBinning.empty()) {
+      std::string errorMsg = std::string("ProtoVolume '") + name +
+                             std::string("' with constituents, but no binning");
+      throw std::runtime_error(errorMsg);
+    }
+
+    // For legacy volumes, check if layers are present
+    bool layersPresent = layerContainer;
```

Comment:
couldn't we use `layerContainer` directly and ditch `layersPresent`?

Reply:
Not really, as `layerPresent` is is an `XAND` of all the possible layer containers, anyhow it is a bit changed now.